packages/peregrine/lib/talons/SignIn/useSignIn.js

```diff
@@ -63,31 +61,36 @@ export const useSignIn = props => {
                 const signInResponse = await signIn({
                     variables: { email, password }
                 });
-                const token = signInResponse.data.generateCustomerToken.token;
-                await setToken(token);
-
-                // Clear all cart/customer data from cache and redux.
-                await clearCartDataFromCache(apolloClient);
-                await clearCustomerDataFromCache(apolloClient);
-                await removeCart();
-
-                // Create and get the customer's cart id.
-                await createCart({
-                    fetchCartId
-                });
-                const destinationCartId = await retrieveCartId();
-
-                // Merge the guest cart into the customer cart.
-                await mergeCarts({
-                    variables: {
-                        destinationCartId,
-                        sourceCartId
-                    }
-                });
-
-                // Ensure old stores are updated with any new data.
-                getUserDetails({ fetchUserDetails });
-                getCartDetails({ fetchCartId, fetchCartDetails });
+                if (signInResponse) {
+                    const token =
+                        signInResponse.data.generateCustomerToken.token;
+                    await setToken(token);
+
+                    // Clear all cart/customer data from cache and redux.
+                    await clearCartDataFromCache(apolloClient);
+                    await clearCustomerDataFromCache(apolloClient);
+                    await removeCart();
+
+                    // Create and get the customer's cart id.
+                    await createCart({
+                        fetchCartId
+                    });
+                    const destinationCartId = await retrieveCartId();
+
+                    // Merge the guest cart into the customer cart.
+                    await mergeCarts({
+                        variables: {
+                            destinationCartId,
+                            sourceCartId
+                        }
+                    });
+
+                    // Ensure old stores are updated with any new data.
+                    getUserDetails({ fetchUserDetails });
+                    getCartDetails({ fetchCartId, fetchCartDetails });
+                } else {
+                    throw new Error('Unable to Signin, please try again.');
```

Comment:
You shouldn't need to do this since the mutation will result in an array, and the awaited sign in should throw.

Reply:
Since I am using `onError` method instead,
```
const signInResponse = await signIn({
    variables: { email, password }
});```
is not throwing an error in case of an error situation. I am not sure if that is expected behavior from Apollo or not because it is not documented. Hence I have changed the working of this function. The function had a try-catch block and I needed it to enter that block manually, hence I did that in the else case.

Reply:
Oh, right, it won't throw, but it will return an error in the mutation response which can be handled by formerror.

Reply:
Ah I misunderstood - regardless, you shouldn't need to do this. The work in #2588 will handle signin errors.